ide/parsing.indexing/src/org/netbeans/modules/parsing/impl/indexing/IndexabilityQuery.java

```diff
@@ -0,0 +1,160 @@
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Add a listener to changes.
+     * @param l a listener to add
+     */
+    public void addChangeListener(ChangeListener l) {
+        synchronized (listeners) {
+            listeners.add(l);
+        }
+    }
+
+    /**
+     * Stop listening to changes.
+     * @param l a listener to remove
+     */
+    public void removeChangeListener(ChangeListener l) {
+        synchronized (listeners) {
+            listeners.remove(l);
+        }
+    }
+    
+    private void fireChange(ChangeEvent event) {
+        assert event != null;
+        ArrayList<ChangeListener> lists;
+        synchronized (listeners) {
+            lists = new ArrayList<>(listeners);
+        }
+        for (ChangeListener listener : lists) {
+            try {
+                listener.stateChanged(event);
+            } catch (RuntimeException x) {
+                Exceptions.printStackTrace(x);
+            }
+        }
+    }
+
+    @SuppressWarnings("ReturnOfCollectionOrArrayField")
+    private synchronized List<IndexabilityQueryImplementation> getIqiInstances() {
+        if (cachedIqiInstances == null) {
+            iqiResult = Lookup.getDefault().lookupResult(IndexabilityQueryImplementation.class);
+            iqiResult.addLookupListener(resultListener);
+            setupChangeListeners(null, new ArrayList<>(iqiResult.allInstances()));
+        }
+        return cachedIqiInstances;
+    }
+
+    @SuppressWarnings("AssignmentToCollectionOrArrayFieldFromParameter")
+    private synchronized void setupChangeListeners(final List<IndexabilityQueryImplementation> oldVqiInstances, final List<IndexabilityQueryImplementation> newVqiInstances) {
+        if (oldVqiInstances != null) {
+            Set<IndexabilityQueryImplementation> removed = new HashSet<>(oldVqiInstances);
+            removed.removeAll(newVqiInstances);
+            for (IndexabilityQueryImplementation vqi : removed) {
+                vqi.removeChangeListener(vqiListener);
+            }
+        }
+
+        Set<IndexabilityQueryImplementation> added = new HashSet<>(newVqiInstances);
+        if (oldVqiInstances != null) {
+            added.removeAll(oldVqiInstances);
+        }
+        for (IndexabilityQueryImplementation vqi : added) {
+            vqi.addChangeListener(vqiListener);
+        }
+
+        cachedIqiInstances = newVqiInstances;
+    }
+
+    private class ResultListener implements LookupListener {
+        @Override
+        public void resultChanged(LookupEvent ev) {
+            setupChangeListeners(cachedIqiInstances, new ArrayList<>(iqiResult.allInstances()));
+            fireChange(new ChangeEvent(this));
+        }
+    }
+
+    private class IqiChangedListener implements ChangeListener {
+        @Override
+        public void stateChanged(ChangeEvent e) {
+            fireChange(e);
```

Comment:
The refired event carries the original `IndexabilityQueryImplementation` as the `getSource()` -- is it intended ?